cpp/arcticdb/async/async_store.hpp

```diff
@@ -280,36 +280,27 @@ class AsyncStore : public Store {
         return res;
     }
 
-    std::vector<VariantKey> batch_read_compressed(
-        std::vector<entity::VariantKey> &&keys,
-        std::vector<ReadContinuation> &&continuations,
+    folly::Future<std::vector<VariantKey>> batch_read_compressed(
+        std::vector<entity::VariantKey>&& keys,
+        std::vector<ReadContinuation>&& continuations,
         const BatchReadArgs & args) override {
-        util::check(keys.size() == continuations.size(),
-                    "keys and continuations must has then same number of elements");
-        std::vector<folly::Future<VariantKey>> batch;
-        batch.reserve(args.batch_size_);
-        std::vector<VariantKey> res;
-        res.reserve(keys.size());
-        for (std::size_t i = 0; i < keys.size(); ++i) {
-            auto &key = keys[i];
-            auto &cont = continuations[i];
-            batch.push_back(
-                async::submit_io_task(ReadCompressedTask(key, library_, storage::ReadKeyOpts{}))
-                    .via(&async::cpu_executor())
-                    .thenValue(SegmentFunctionTask{std::move(cont)}));
 
-            if(batch.size() == args.batch_size_) {
-                auto vec = folly::collect(batch).get();
-                res.insert(end(res), std::make_move_iterator(std::begin(vec)), std::make_move_iterator(std::end(vec)));
-                batch.clear();
-            }
-        }
+        util::check(!keys.empty(), "Unexpected empty keys in batch_read_compressed");
 
-        if(!batch.empty()) {
-            auto vec = folly::collect(batch).get();
-            res.insert(end(res), std::make_move_iterator(std::begin(vec)), std::make_move_iterator(std::end(vec)));
+        auto key_seg_futs = folly::window(keys, [*this] (auto&& key) {
+                return async::submit_io_task(ReadCompressedTask(std::move(key), library_, storage::ReadKeyOpts{}));
+            }, args.batch_size_);
+
+        util::check(key_seg_futs.size() == keys.size(), "Size mismatch in batch_read_compressed: {} != {}", key_seg_futs.size(), keys.size());
+        std::vector<folly::Future<VariantKey>> result;
+        result.reserve(key_seg_futs.size());
```

Comment:
As previously discussed, I am not requesting any changes. However, it's important to note that there is a subtle difference in performance due to the fact that the folly window method is now only applied to the ReadCompressedTask, rather than both the ReadCompressedTask and SegmentFunctionTask. While we are still throttling, there is no control over the number of threads executing the SegmentFunctionTask.

Reply:
That's fine in my opinion, because we are gating the number of things that can enter the pipeline